File: tools/parity/ts/dumpFormat.ts
/**
 * Shared on-disk dump format for TS <-> Kotlin numeric parity.
 *
 * Two files per pipeline stage:
 *
 *   <stage>.f64   raw IEEE-754 binary64, little-endian, row-major:
 *                 value(point i, field f) is at byte offset (i * fieldCount + f) * 8
 *   <stage>.json  header: stage name, point count, field order, path aggregates
 *
 * Binary (rather than JSON Lines) is deliberate: it round-trips the exact bit pattern of
 * every double with no decimal formatting in the loop, and keeps a 50 k-point stage at
 * ~14 MB instead of ~150 MB. `compare.py` reads both sides with the same struct layout,
 * so a difference in the report is always a difference in the computed value.
 *
 * The Kotlin writer (`DumpFormat.kt`) MUST stay byte-compatible with this file.
 */
import { closeSync, mkdirSync, openSync, writeFileSync, writeSync } from 'node:fs';
import { dirname, join } from 'node:path';

/** Aggregates mirrored from `Path`'s cached statistics (Kotlin `Path` property names). */
export interface DumpHeader {
    stage: string;
    index: number;
    size: number;
    fieldCount: number;
    fields: string[];
    totalDistance: number;
    durationMs: number;
    elevationGain: number;
    elevationLoss: number;
    minElevation: number;
    maxElevation: number;
    /** Epoch offset subtracted from every timestamp at parse time (see runner). */
    timeOrigin: number;
}

/**
 * Write one stage. `read(i, f)` returns the raw stored double for point `i`, field `f`,
 * where `f` indexes `fields` (== PointField ordinal on both sides).
 */
export function writeStage(
    outDir: string,
    header: DumpHeader,
    read: (pointIndex: number, fieldIndex: number) => number
): void {
    mkdirSync(outDir, { recursive: true });

    const stem = `${String(header.index).padStart(2, '0')}-${header.stage}`;
    const binPath = join(outDir, `${stem}.f64`);
    const jsonPath = join(outDir, `${stem}.json`);

    const { size, fieldCount } = header;
    // Chunked write: one 1 MiB-ish buffer reused, so a 100 k-point stage never allocates
    // a 300 MB Buffer.
    const pointsPerChunk = Math.max(1, Math.floor(4096 / fieldCount) * 8);
    const chunk = Buffer.allocUnsafe(pointsPerChunk * fieldCount * 8);

    mkdirSync(dirname(binPath), { recursive: true });
    const fd = openSync(binPath, 'w');
    try {
        for (let start = 0; start < size; start += pointsPerChunk) {
            const count = Math.min(pointsPerChunk, size - start);
            let off = 0;
            for (let i = start; i < start + count; i++) {
                for (let f = 0; f < fieldCount; f++) {
                    chunk.writeDoubleLE(read(i, f), off);
                    off += 8;
                }
            }
            writeSync(fd, chunk, 0, off);
        }
    } finally {
        closeSync(fd);
    }

    writeFileSync(jsonPath, `${JSON.stringify(header, null, 2)}\n`);
}

File: tools/parity/src/main/kotlin/io/github/glandais/parity/DumpFormat.kt
package io.github.glandais.parity

import io.github.glandais.engine.path.Path
import io.github.glandais.engine.path.PointField
import java.io.BufferedOutputStream
import java.io.File
import java.nio.ByteBuffer
import java.nio.ByteOrder

/**
 * Shared on-disk dump format for TS <-> Kotlin numeric parity.
 *
 * Byte-compatible with `tools/parity/ts/dumpFormat.ts` — see that file for the format
 * description. Two files per stage :
 *
 * - `<nn>-<stage>.f64` : raw IEEE-754 binary64, little-endian, row-major,
 *   `value(i, f)` at byte offset `(i * fieldCount + f) * 8`.
 * - `<nn>-<stage>.json` : header (stage, size, field order, path aggregates).
 *
 * Binary rather than text so no decimal formatting sits between the computed double and
 * the comparison — a difference in the report is always a difference in the value.
 */
object DumpFormat {
    /** Field order == [PointField] ordinal, identical to the TS `POINT_FIELDS` order. */
    val FIELD_NAMES: List<String> = PointField.entries.map { it.prop }

    /**
     * Write one pipeline stage of [path] into [outDir].
     *
     * @param index zero-based stage counter (prefixes the filename so stages sort in order)
     * @param timeOrigin epoch offset subtracted from every timestamp at parse time
     */
    fun writeStage(
        outDir: File,
        index: Int,
        stage: String,
        path: Path,
        timeOrigin: Double,
    ) {
        outDir.mkdirs()
        val stem = "%02d-%s".format(index, stage)
        val fieldCount = PointField.COUNT

        // Chunked so a 100 k-point stage never allocates a 300 MB array.
        val pointsPerChunk = 4096
        val buffer =
            ByteBuffer
                .allocate(pointsPerChunk * fieldCount * 8)
                .order(ByteOrder.LITTLE_ENDIAN)

        BufferedOutputStream(File(outDir, "$stem.f64").outputStream()).use { out ->
            var start = 0
            while (start < path.size) {
                val count = minOf(pointsPerChunk, path.size - start)
                buffer.clear()
                for (i in start until start + count) {
                    for (field in PointField.entries) {
                        buffer.putDouble(path.get(i, field))
                    }
                }
                out.write(buffer.array(), 0, count * fieldCount * 8)
                start += count
            }
        }

        val durationMs =
            if (path.size >= 2) path.time(path.size - 1) - path.time(0) else 0.0
        val fieldsJson = FIELD_NAMES.joinToString(",\n    ") { "\"$it\"" }
        File(outDir, "$stem.json").writeText(
            """
            {
              "stage": "$stage",
              "index": $index,
              "size": ${path.size},
              "fieldCount": $fieldCount,
              "fields": [
                $fieldsJson
              ],
              "totalDistance": ${json(path.totalDistance)},
              "durationMs": ${json(durationMs)},
              "elevationGain": ${json(path.elevationGain)},
              "elevationLoss": ${json(path.elevationLoss)},
              "minElevation": ${json(if (path.size == 0) 0.0 else path.minElevation)},
              "maxElevation": ${json(if (path.size == 0) 0.0 else path.maxElevation)},
              "timeOrigin": ${json(timeOrigin)}
            }
            """.trimIndent() + "\n",
        )
    }

    /**
     * JSON-safe rendering of a double. `Double.toString` is round-trip exact on the JVM, but
     * JSON has no literal for the non-finite values, so those degrade to `null` rather than
     * producing a file no parser will accept.
     */
    fun json(v: Double): String = if (v.isFinite()) v.toString() else "null"
}
